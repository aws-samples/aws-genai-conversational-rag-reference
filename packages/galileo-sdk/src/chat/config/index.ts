/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import type {
  ChatEngineConfig,
  ChatEngineChainConfig,
  ChatEngineLLMConfig,
  ChatEngineSearchConfig,
  ChatEngineMemoryConfig,
} from 'api-typescript-runtime';
import { BaseLanguageModel } from 'langchain/base_language';
import { PromptTemplate } from 'langchain/prompts';
import { difference } from 'lodash';
import { ChainType } from '../../schema/index.js';
import { mergeConfig } from '../../utils/merge.js';
import { ChatEngineContext, ResolvedLLM } from '../context.js';

export { ChatEngineConfig, ChatEngineChainConfig, ChatEngineLLMConfig, ChatEngineSearchConfig, ChatEngineMemoryConfig };

export interface ResolvedLLMChainConfig {
  readonly llm: BaseLanguageModel;
  readonly prompt: PromptTemplate;
}

export interface ResolvedChatEngineConfig {
  /** Indicates if config is the root, and should not be merged with ancestors */
  readonly root?: boolean;
  /** LLM to use for chains, unless overridden by the chain */
  readonly llm: BaseLanguageModel;
  /** Classify chain config, if undefined no classification will be performed */
  readonly classifyChain?: ResolvedLLMChainConfig;
  /** Question/Answer chain config */
  readonly qaChain: ResolvedLLMChainConfig;
  /** Condense question chain (standalone question generator) */
  readonly condenseQuestionChain: ResolvedLLMChainConfig;

  readonly search?: ChatEngineSearchConfig;

  readonly memory?: ChatEngineMemoryConfig;
}

export function mergeUnresolvedChatEngineConfig(...configs: Partial<ChatEngineConfig>[]): ChatEngineConfig {
  return mergeConfig(...(configs as any)) as ChatEngineConfig;
}

function extractPrivilegedKeys(obj: any, allowed: string[], prefix?: string): string[] {
  if (obj && typeof obj === 'object') {
    const privileged = difference(Object.keys(obj), allowed);
    if (privileged.length) {
      return prefix ? privileged.map((v) => `${prefix}.${v}`) : privileged;
    }
  }
  return [];
}

export const UNPRIVILEGED_KEYS = {
  MODEL: ['uuid'],
  SEARCH: ['filter', 'limit', 'scoreThreshold'],
};

export function extractPrivilegedChatEngineConfigKeys(config: Partial<ChatEngineConfig>): string[] {
  const privileged: string[] = [];
  privileged.push(...extractPrivilegedKeys(config.llm?.model, UNPRIVILEGED_KEYS.MODEL, 'llm.model'));

  Object.values(ChainType).forEach((type) => {
    const key = getChainConfigKeyByType(type);
    const chainConfig = config[key] as ChatEngineChainConfig | undefined;
    privileged.push(...extractPrivilegedKeys(chainConfig?.llm?.model, UNPRIVILEGED_KEYS.MODEL, `${key}.llm.model`));
  });

  privileged.push(...extractPrivilegedKeys(config.search, UNPRIVILEGED_KEYS.SEARCH, 'search'));

  return privileged;
}

export function assertNonPrivilegedChatEngineConfig(config: Partial<ChatEngineChainConfig>): void {
  const privileged = extractPrivilegedChatEngineConfigKeys(config);

  if (privileged.length) {
    throw new Error(`ChatEngineConfig contains the following privileged properties: ${privileged.join(', ')}`);
  }
}

async function resolveLLM(
  llmConfig?: ChatEngineLLMConfig,
  defaultValues?: ResolvedLLM,
  options?: { verbose?: boolean },
): Promise<ResolvedLLM> {
  let modelInfo = defaultValues?.modelInfo;
  if (llmConfig?.model) {
    modelInfo = await ChatEngineContext.resolveModelInfo(llmConfig.model);
  }

  return ChatEngineContext.resolveLLM(modelInfo, { ...llmConfig, ...options });
}

export async function resolveChatEngineConfig(
  config: ChatEngineConfig,
  options?: { verbose?: boolean },
): Promise<ResolvedChatEngineConfig> {
  const defaultLLM = await resolveLLM(config.llm, undefined, options);
  const qaLLM = await resolveLLM(config.qaChain?.llm, defaultLLM, options);
  const condensedLLM = await resolveLLM(config.condenseQuestionChain?.llm, defaultLLM, options);
  const classifyLLM = config.classifyChain?.enabled
    ? await resolveLLM(config.classifyChain?.llm, defaultLLM, options)
    : undefined;

  return {
    ...config,
    llm: defaultLLM.llm,
    qaChain: {
      llm: qaLLM.llm,
      prompt: await ChatEngineContext.resolvePromptTemplate(
        ChainType.QA,
        qaLLM.adapter.prompt?.chat?.QA,
        config.qaChain?.prompt,
      ),
    },
    condenseQuestionChain: {
      llm: condensedLLM.llm,
      prompt: await ChatEngineContext.resolvePromptTemplate(
        ChainType.CONDENSE_QUESTION,
        condensedLLM.adapter.prompt?.chat?.CONDENSE_QUESTION,
        config.condenseQuestionChain?.prompt,
      ),
    },
    classifyChain:
      config.classifyChain?.enabled && classifyLLM
        ? {
            llm: classifyLLM.llm,
            prompt: await ChatEngineContext.resolvePromptTemplate(
              ChainType.CLASSIFY,
              classifyLLM.adapter.prompt?.chat?.CLASSIFY,
              config.classifyChain?.prompt,
            ),
          }
        : undefined,
  };
}

export function getChainConfigKeyByType(type: ChainType) {
  switch (type) {
    case ChainType.QA:
      return 'qaChain';
    case ChainType.CONDENSE_QUESTION:
      return 'condenseQuestionChain';
    case ChainType.CLASSIFY:
      return 'classifyChain';
  }
}

export function getChainConfigByType<T extends ChatEngineConfig | ResolvedChatEngineConfig>(
  type: ChainType,
  config?: T,
) {
  return config && config[getChainConfigKeyByType(type)];
}
