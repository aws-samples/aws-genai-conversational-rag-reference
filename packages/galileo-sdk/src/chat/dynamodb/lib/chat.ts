/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import {
  DynamoDBDocumentClient,
  PutCommand,
  QueryCommandInput,
  DeleteCommand,
  UpdateCommand,
} from '@aws-sdk/lib-dynamodb';
import { v4 as uuidv4 } from 'uuid';
import { getAllChatMessageIds } from './messages.js';
import { listChatMessageSources } from './sources.js';
import {
  getChatKey,
  getChatsByTimeKey,
  DDBChat,
  bulkDelete,
  Keys,
  DDBUpdateOutput,
  getAllByPagination,
} from './util.js';

export async function createChat(
  documentClient: DynamoDBDocumentClient,
  tableName: string,
  userId: string,
  title: string,
) {
  const newChatSessionId = uuidv4();

  const timestamp = Date.now();
  const keys = getChatKey(userId, newChatSessionId);
  const gsiKeys = getChatsByTimeKey(userId, `${timestamp}`);
  const chat: DDBChat = {
    chatId: newChatSessionId,
    title: title,
    createdAt: timestamp,
    userId: userId,
    ...keys,
    ...gsiKeys,
    entity: 'CHAT',
  };

  const command = new PutCommand({
    TableName: tableName,
    Item: chat,
    ReturnValues: 'NONE',
  });

  const response = await documentClient.send(command);

  return {
    response,
    chat,
  };
}

export async function listAllChatsByTime(
  documentClient: DynamoDBDocumentClient,
  tableName: string,
  indexName: string,
  userId: string,
  asc = false,
) {
  const keys = getChatsByTimeKey(userId);

  const commandInput: QueryCommandInput = {
    TableName: tableName,
    IndexName: indexName,
    KeyConditionExpression: 'GSI1PK = :PK',
    ExpressionAttributeValues: {
      ':PK': keys.GSI1PK,
    },
    ScanIndexForward: asc,
  };

  // We've decided to return all chats for a user and then filter and paginate client side
  return getAllByPagination<DDBChat>(documentClient, commandInput);
}

export async function updateChat(
  documentClient: DynamoDBDocumentClient,
  tableName: string,
  userId: string,
  chatId: string,
  title: string,
) {
  const keys = getChatKey(userId, chatId);

  const command = new UpdateCommand({
    TableName: tableName,
    Key: keys,
    ConditionExpression: 'attribute_exists(PK) and attribute_exists(SK)',
    UpdateExpression: 'set title = :title',
    ExpressionAttributeValues: {
      ':title': title,
    },
    ReturnValues: 'ALL_NEW',
  });

  const result = (await documentClient.send(command)) as DDBUpdateOutput<DDBChat>;
  return result.Attributes;
}

async function deleteChat(documentClient: DynamoDBDocumentClient, tableName: string, userId: string, chatId: string) {
  const keys = getChatKey(userId, chatId);

  const command = new DeleteCommand({
    TableName: tableName,
    Key: keys,
    ReturnValues: 'ALL_OLD',
  });

  return documentClient.send(command);
}

export async function deleteChatAndMessages(
  documentClient: DynamoDBDocumentClient,
  tableName: string,
  indexName: string,
  userId: string,
  chatId: string,
) {
  // For now we'll perform all the delete actions in a single lambda
  // We might want to consider creating a job queue for deleting the chatMessages async
  const chatMessageKeys = await getAllChatMessageIds(documentClient, tableName, indexName, userId, chatId);

  let chatMessageSourceKeys: Keys[] = [];
  for (const messageKeys of chatMessageKeys) {
    const response = await listChatMessageSources(documentClient, tableName, userId, messageKeys.messageId);
    if (response) {
      chatMessageSourceKeys = [...chatMessageSourceKeys, ...response];
    }
  }

  const deleteKeys = [...chatMessageKeys, ...chatMessageSourceKeys].map((k) => ({ PK: k.PK, SK: k.SK }));

  console.log(`found ${chatMessageKeys.length} messages`);
  console.log(`total records to delete: ${deleteKeys.length}`);

  if (chatMessageKeys.length > 0) {
    await bulkDelete(documentClient, tableName, deleteKeys);
  }
  await deleteChat(documentClient, tableName, userId, chatId);
}
