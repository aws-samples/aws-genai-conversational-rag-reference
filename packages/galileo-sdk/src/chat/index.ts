/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import '../langchain/patch.js';
export * as dynamodb from './dynamodb/index.js';
export * as chain from './chain.js';
export * as context from './context.js';
export * as memory from './memory.js';

export * from './engine.js';
