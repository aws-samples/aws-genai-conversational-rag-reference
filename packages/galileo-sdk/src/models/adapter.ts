/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import { BaseSageMakerContentHandler } from 'langchain/llms/sagemaker_endpoint';
import { set, get, isEmpty } from 'lodash';
import { getLogger } from '../common/index.js';
import { BaseChatTemplatePartials, ChatCondenseQuestionPromptRuntime, ChatQuestionAnswerPromptRuntime } from '../prompt/templates/chat/index.js';

const logger = getLogger('models/adapter');

/**
 * @struct
 */
export interface IContentHandlerAdapterInput {
  /**
   * Path in input object to define the prompt value.
   * @see https://lodash.com/docs/#set
   * @default "text_inputs"
   */
  readonly promptKey?: string;
  /**
   * Path in the input object to define the model kwargs.
   * @see https://lodash.com/docs/#set
   * @default undefined - Kwargs will be spread on root
   */
  readonly modelKwargsKey?: string;
  // readonly prepend?: string;
  // readonly append?: string;
  /**
   * @default "utf-8"
   */
  readonly encoding?: BufferEncoding;
}
/**
 * @struct
 */
export interface IContentHandlerAdapterOutput {
  /**
   * Path of response object to extract as output.
   * @see https://lodash.com/docs/#get
   * @default `[0].generated_text`
   */
  readonly jsonpath?: string;
  /**
   * @default "utf-8"
   */
  readonly encoding?: BufferEncoding;
}
/**
 * @struct
 */
export interface IContentHandlerAdapter {
  readonly contentType?: string;
  readonly accepts?: string;
  readonly input?: IContentHandlerAdapterInput;
  readonly output?: IContentHandlerAdapterOutput;
}

/**
 * @struct
 */
export interface IChatPromptAdapter {
  readonly base?: Partial<BaseChatTemplatePartials>;
  readonly questionAnswer?: ChatQuestionAnswerPromptRuntime;
  readonly condenseQuestion?: ChatCondenseQuestionPromptRuntime;
}
/**
 * @struct
 */
export interface IPromptAdapter {
  readonly chat?: IChatPromptAdapter;
}

export class PromptAdapter implements IPromptAdapter {

  readonly chat?: IChatPromptAdapter;

  constructor(adapter?: IPromptAdapter) {
    this.chat = {
      base: adapter?.chat?.base,
      condenseQuestion: adapter?.chat?.base || adapter?.chat?.condenseQuestion ? {
        ...adapter.chat.condenseQuestion,
        templatePartials: {
          ...adapter.chat.base,
          ...adapter.chat.condenseQuestion?.templatePartials,
        },
      } : undefined,
      questionAnswer: adapter?.chat?.base || adapter?.chat?.questionAnswer ? {
        ...adapter.chat.questionAnswer,
        templatePartials: {
          ...adapter.chat.base,
          ...adapter.chat.questionAnswer?.templatePartials,
        },
      } : undefined,
    };
  }
}

/**
 * @struct
 */
export interface IModelAdapter {
  readonly prompt?: IPromptAdapter;
  readonly contentHandler?: IContentHandlerAdapter;
}

export const DEFAULT_OUTPUT_JSONPATH = '[0].generated_text';

export class AdaptedContentHandler extends BaseSageMakerContentHandler<string, string> {
  readonly contentType: string;
  readonly accepts: string;

  protected readonly promptKey: string;
  protected readonly modelKwargsKey?: string;
  protected readonly inputEncoding: BufferEncoding;

  protected readonly outputJsonpath: string;
  protected readonly outputEncoding: BufferEncoding;

  constructor(adapter?: IContentHandlerAdapter) {
    super();

    this.accepts = adapter?.accepts ?? 'application/json';
    this.contentType = adapter?.contentType ?? 'application/json';

    this.promptKey = adapter?.input?.promptKey ?? 'text_inputs';
    this.modelKwargsKey = adapter?.input?.modelKwargsKey;
    this.inputEncoding = adapter?.input?.encoding ?? 'utf-8';

    this.outputJsonpath = adapter?.output?.jsonpath ?? DEFAULT_OUTPUT_JSONPATH;
    this.outputEncoding = adapter?.output?.encoding ?? 'utf-8';

    if (this.accepts !== 'application/json' || this.contentType != 'application/json') {
      throw new Error(`Only "application/json" is supported at this time: ${this.accepts} / ${this.contentType} not implemented`);
    }
  }

  async transformInput(prompt: string, modelKwargs: Record<string, unknown>): Promise<Uint8Array> {
    const input = {};

    set(input, this.promptKey, prompt);

    if (this.modelKwargsKey) {
      set(input, this.modelKwargsKey, modelKwargs);
    } else {
      // spread kwargs if not key
      Object.assign(input, modelKwargs);
    }

    logger.debug('AdaptedContentHandler#transformInput:', { prompt, modelKwargs, input });

    return Buffer.from(JSON.stringify(input), this.inputEncoding);
  }

  async transformOutput(output: Uint8Array): Promise<string> {
    const response = JSON.parse(Buffer.from(output).toString(this.outputEncoding));
    // extract the path from json
    const result = get(response, this.outputJsonpath);
    logger.debug('AdaptedContentHandler#transformOutput:', { response, result });
    return result;
  }
}

export class ModelAdapter {
  readonly prompt?: IPromptAdapter;

  readonly contentHandler: AdaptedContentHandler;

  readonly isDefault: boolean;

  constructor(config?: IModelAdapter) {
    this.isDefault = config == null || isEmpty(config);

    this.prompt = new PromptAdapter(config?.prompt);

    this.contentHandler = new AdaptedContentHandler(config?.contentHandler);
  }
}
