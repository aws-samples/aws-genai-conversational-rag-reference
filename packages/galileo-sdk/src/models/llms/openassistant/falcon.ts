/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import { CHAT_QUESTION_ANSWER_TEMPLATE } from '../../../prompt/index.js';
import type { IModelAdapter } from '../../adapter.js';
import { DEFAULT_MAX_NEW_TOKENS } from '../../constants.js';
import type { Kwargs } from '../../types.js';

export const FALCON_ADAPTER: IModelAdapter = {
  prompt: {
    chat: {
      base: {
        AIMessage: '<|assistant|>{{content}}<|endoftext|>',
        HumanMessage: '<|prompter|>{{content}}<|endoftext|>',
        SystemMessage: '<|system|>{{content}}<|endoftext|>',
      },
      questionAnswer: {
        template: CHAT_QUESTION_ANSWER_TEMPLATE.replace(
          /Question: {{question}}.*$/i,
          'Question: <|prompter|>{{question}}<|endoftext|>\n\nAnswer: <|assistant|>',
        ),
      },
    },
  },
  contentHandler: {
    input: {
      promptKey: 'inputs',
      modelKwargsKey: 'parameters',
    },
    output: {
      jsonpath: '[0].generated_text',
    },
  },
};

export const FALCON_MODEL_KWARGS: Kwargs = {
  do_sample: true,
  top_p: 0.6,
  temperature: 0.001,
  top_k: 30,
  max_new_tokens: DEFAULT_MAX_NEW_TOKENS,
  repetition_penalty: 1.03,
  stop: ['<|endoftext|>'],
  return_full_text: false,
  typical_p: 0.2,
};
