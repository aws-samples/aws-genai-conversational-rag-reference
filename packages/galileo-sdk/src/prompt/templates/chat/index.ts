/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
export * from './base.js';
export * from './condense-question.js';
export * from './question-answer.js';
