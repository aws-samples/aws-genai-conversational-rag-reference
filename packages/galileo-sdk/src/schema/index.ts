/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
export enum ChainType {
  QA = 'QA',
  CONDENSE_QUESTION = 'CONDENSE_QUESTION',
  CLASSIFY = 'CLASSIFY',
}
