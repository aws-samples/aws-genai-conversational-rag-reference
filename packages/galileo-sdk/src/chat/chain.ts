/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import { BaseLanguageModel } from 'langchain/base_language';
import { CallbackManagerForChainRun } from 'langchain/callbacks';
import { BaseChain, ChainInputs, LLMChain, loadQAChain, QAChainParams } from 'langchain/chains';
import { PromptTemplate } from 'langchain/prompts';
// import { SerializedChatVectorDBQAChain } from "./serde.js";
import {
  ChainValues,
} from 'langchain/schema';
import { BaseRetriever } from 'langchain/schema/retriever';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type LoadValues = Record<string, any>;

export interface ChatEngineChainInput extends ChainInputs {
  retriever: BaseRetriever;
  combineDocumentsChain: BaseChain;
  questionGeneratorChain: LLMChain;
  returnSourceDocuments?: boolean;
  inputKey?: string;
}

export class ChatEngineChain extends BaseChain implements ChatEngineChainInput {
  static lc_name() {
    return 'ChatEngineChain';
  }

  /**
   * Static method to create a new ChatEngineChain from a
   * BaseLanguageModel and a BaseRetriever.
   * @param llm {@link BaseLanguageModel} instance used to generate a new question.
   * @param retriever {@link BaseRetriever} instance used to retrieve relevant documents.
   * @param options.returnSourceDocuments Whether to return source documents in the final output
   * @param options.questionGeneratorChainOptions Options to initialize the standalone question generation chain used as the first internal step
   * @param options.qaChainOptions {@link QAChainParams} used to initialize the QA chain used as the second internal step
   * @returns A new instance of ChatEngineChain.
   */
  static fromLLM(
    llm: BaseLanguageModel,
    retriever: BaseRetriever,
    options: {
      returnSourceDocuments?: boolean;
      questionGenerator: {
        llm?: BaseLanguageModel;
        prompt: PromptTemplate;
      };
      qaChainOptions: QAChainParams;
    } & Omit<
    ChatEngineChainInput,
    'retriever' | 'combineDocumentsChain' | 'questionGeneratorChain'
    >,
  ): ChatEngineChain {
    const {
      qaChainOptions,
      questionGenerator,
      verbose,
      ...rest
    } = options;

    const qaChain = loadQAChain(llm, qaChainOptions);

    const questionGeneratorChain = new LLMChain({
      prompt: questionGenerator.prompt,
      llm: questionGenerator.llm ?? llm,
      verbose,
    });

    const instance = new this({
      retriever,
      combineDocumentsChain: qaChain,
      questionGeneratorChain,
      verbose,
      ...rest,
    });
    return instance;
  }

  inputKey = 'question';

  chatHistoryKey = 'chat_history';

  get inputKeys() {
    return [this.inputKey, this.chatHistoryKey];
  }

  get outputKeys() {
    return this.combineDocumentsChain.outputKeys.concat(
      this.returnSourceDocuments ? ['sourceDocuments'] : [],
    );
  }

  retriever: BaseRetriever;

  combineDocumentsChain: BaseChain;

  questionGeneratorChain: LLMChain;

  returnSourceDocuments = false;

  constructor(fields: ChatEngineChainInput) {
    super(fields);
    this.retriever = fields.retriever;
    this.combineDocumentsChain = fields.combineDocumentsChain;
    this.questionGeneratorChain = fields.questionGeneratorChain;
    this.inputKey = fields.inputKey ?? this.inputKey;
    this.returnSourceDocuments =
      fields.returnSourceDocuments ?? this.returnSourceDocuments;
  }

  /** @ignore */
  async _call(
    values: ChainValues,
    runManager?: CallbackManagerForChainRun,
  ): Promise<ChainValues> {
    if (!(this.inputKey in values)) {
      throw new Error(`Question key ${this.inputKey} not found.`);
    }
    if (!(this.chatHistoryKey in values)) {
      throw new Error(`Chat history key ${this.chatHistoryKey} not found.`);
    }
    const question: string = values[this.inputKey];
    const chatHistory = values[this.chatHistoryKey];

    let newQuestion = question;
    if (chatHistory.length > 0) {
      const result = await this.questionGeneratorChain.call(
        {
          question,
          chat_history: chatHistory,
        },
        runManager?.getChild('question_generator'),
      );
      const keys = Object.keys(result);
      if (keys.length === 1) {
        newQuestion = result[keys[0]];
      } else {
        throw new Error(
          'Return from llm chain has multiple values, only single values supported.',
        );
      }
    }
    const docs = await this.retriever.getRelevantDocuments(
      newQuestion,
      runManager?.getChild('retriever'),
    );
    const inputs = {
      question: newQuestion,
      input_documents: docs,
      chat_history: chatHistory,
    };
    const result = await this.combineDocumentsChain.call(
      inputs,
      runManager?.getChild('combine_documents'),
    );
    if (this.returnSourceDocuments) {
      return {
        ...result,
        sourceDocuments: docs,
      };
    }
    return result;
  }

  _chainType(): string {
    return 'conversational_retrieval_chain';
  }

  // static async deserialize(
  //   _data: SerializedChatVectorDBQAChain,
  //   _values: LoadValues
  // ): Promise<ChatEngineChain> {
  //   throw new Error("Not implemented.");
  // }

  // serialize(): SerializedChatVectorDBQAChain {
  //   throw new Error("Not implemented.");
  // }
}
