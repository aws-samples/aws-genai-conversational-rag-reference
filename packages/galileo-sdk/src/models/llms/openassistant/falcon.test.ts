/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
// @ts-ignore - .test files are ignored
import type {} from '@types/jest';
import { AIMessage, HumanMessage, SystemMessage } from 'langchain/schema';
import { FALCON_ADAPTER } from './falcon';
import { ChatCondenseQuestionPromptTemplate, ChatQuestionAnswerPromptTemplate } from '../../../prompt/templates/chat';
import { ModelAdapter } from '../../adapter';

describe('models/llms/openassistant/falcon', () => {
  describe('adapter', () => {
    const adapter = new ModelAdapter(FALCON_ADAPTER);

    test('should render qa prompt', async () => {
      const template = new ChatQuestionAnswerPromptTemplate({
        ...adapter.prompt?.chat?.questionAnswer,
      });
      expect(
        await template.format({
          question: 'THE QUESTION',
          domain: 'DOMAIN',
          context: 'THE CORPUS',
        }),
      ).toMatchSnapshot();
    });

    test('should render condense prompt', async () => {
      const template = new ChatCondenseQuestionPromptTemplate({
        ...adapter.prompt?.chat?.condenseQuestion,
      });
      expect(
        await template.format({
          question: 'THE QUESTION',
          chat_history: [new SystemMessage('A system message'), new HumanMessage('Hi'), new AIMessage('Hello!')],
        }),
      ).toMatchSnapshot();
    });
  });
});
