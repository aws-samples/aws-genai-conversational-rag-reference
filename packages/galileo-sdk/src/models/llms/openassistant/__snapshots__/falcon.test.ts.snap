// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models/llms/openassistant/falcon adapter should render condense prompt 1`] = `
"Given the following conversational dialog delimited by ''', and the "Followup Question" below, rephrase the "Followup Question" to be a concise standalone question in its original language. Without answering the question, return only the standalone question.

Dialog: '''
<|system|>A system message<|endoftext|>
<|prompter|>Hi<|endoftext|>
<|assistant|>Hello!<|endoftext|>
'''

Followup Question: THE QUESTION
Standalone Question: "
`;

exports[`models/llms/openassistant/falcon adapter should render qa prompt 1`] = `
"You are a research assistant in the "DOMAIN" domain. Based on the following rules and provided corpus delimited by ''', answer the question. 

Rules:
1. only use knowledge from the provided corpus to answer the question
2. always be truthful, honest, unbiased, and unharmful
3. be concise, do not repeat the question or yourself in the answer

Corpus: '''
THE CORPUS
'''

Question: <|prompter|>THE QUESTION<|endoftext|>

Answer: <|assistant|>"
`;
